File: src/main/java/com/linecorp/armeria/server/logging/LoggingService.java
/*
 * Copyright 2016 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.linecorp.armeria.server.logging;

import static java.util.Objects.requireNonNull;

import com.linecorp.armeria.common.Request;
import com.linecorp.armeria.common.RequestContext;
import com.linecorp.armeria.common.Response;
import com.linecorp.armeria.common.logging.LogLevel;
import com.linecorp.armeria.common.logging.MessageLogConsumer;
import com.linecorp.armeria.common.logging.RequestLog;
import com.linecorp.armeria.common.logging.ResponseLog;
import com.linecorp.armeria.server.Service;
import com.linecorp.armeria.server.ServiceRequestContext;

/**
 * Decorates a {@link Service} to log {@link Request}s and {@link Response}s.
 *
 * @param <I> the {@link Request} type
 * @param <O> the {@link Response} type
 */
public class LoggingService<I extends Request, O extends Response> extends LogCollectingService<I, O> {

    /**
     * Creates a new instance that logs {@link Request}s and {@link Response}s at {@link LogLevel#INFO}.
     */
    public LoggingService(Service<? super I, ? extends O> delegate) {
        this(delegate, LogLevel.INFO);
    }

    /**
     * Creates a new instance that logs {@link Request}s and {@link Response}s at the specified
     * {@link LogLevel}.
     */
    public LoggingService(Service<? super I, ? extends O> delegate, LogLevel level) {
        super(delegate, new LoggingConsumer(level));
    }

    private static final class LoggingConsumer implements MessageLogConsumer {

        private static final String REQUEST_FORMAT = "Request: {}";
        private static final String RESPONSE_FORMAT = "Response: {}";

        private final LogLevel level;

        LoggingConsumer(LogLevel level) {
            this.level = requireNonNull(level, "level");
        }

        @Override
        public void onRequest(RequestContext ctx, RequestLog req) {
            level.log(((ServiceRequestContext) ctx).logger(), REQUEST_FORMAT, req);
        }

        @Override
        public void onResponse(RequestContext ctx, ResponseLog res) {
            level.log(((ServiceRequestContext) ctx).logger(), RESPONSE_FORMAT, res);
        }
    }
}

File: kotlin/src/main/kotlin/com/linecorp/armeria/server/kotlin/logging/LoggingService.kt
/*
 * Copyright 2021 LINE Corporation
 *
 * LINE Corporation licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

package com.linecorp.armeria.server.kotlin.logging

import com.linecorp.armeria.server.HttpService
import com.linecorp.armeria.server.ServerBuilder
import com.linecorp.armeria.server.logging.LoggingService
import com.linecorp.armeria.server.logging.LoggingServiceBuilder

/**
 * Decorates this [HttpService] using a [LoggingService].
 */
fun HttpService.logging(): LoggingService =
    decorate(LoggingService.newDecorator())

/**
 * Decorates this [HttpService] using a [LoggingService] configured by the given [block].
 */
fun HttpService.logging(block: LoggingServiceBuilder.() -> Unit): LoggingService =
    decorate(LoggingService.builder().apply(block).newDecorator())

/**
 * Decorates all [HttpService]s using a [LoggingService].
 */
fun ServerBuilder.logging(): ServerBuilder =
    decorator(LoggingService.newDecorator())

/**
 * Decorates all [HttpService]s using a [LoggingService] configured by the given [block].
 */
fun ServerBuilder.logging(block: LoggingServiceBuilder.() -> Unit): ServerBuilder =
    decorator(LoggingService.builder().apply(block).newDecorator())
